### PR TITLE
Fix derivative of LogSoftMax (Resolves #3468)

### LINKED ISSUE #3468: Incorrect Derivative of LogSoftMax layer
**Description**: The derivative used in the Backward method of the LogSoftMax layer is incorrect. This issue seemed to be noticed in the past but was closed, possibly incorrectly(?) (See #2364  )

The derivative calculated using LogSoftMax should be equal to derivative of logarithm applied on the Softmax layer. But, the derivative using both are significantly different. 

**Code to reproduce**
```
#include <mlpack.hpp>

using namespace mlpack;
using namespace arma;

int main(){
  int n = 5, m = 1;
  mat input = arma::mat("-0.6871; 0.7898; 0.2011; 0.0949; -0.0550");
  input.print("Input");
  mat output(n, m), delta(n, m); 

  LogSoftMax lsm;
  lsm.Forward(input, output);
  mat gy = ones(n, m); 
  lsm.Backward(output, gy, delta);

  mat outputT(n, m), delta2(n, m); 

  Softmax sm; 
  sm.Forward(input, outputT);
  mat output2 = arma::log(outputT);
  //Derivative of log function. 
  gy = arma::pow(outputT, -1);
  
  //Derivative of the Softmax layer 
  sm.Backward(outputT, gy, delta2);

  output.print("LogSoftMax Output:");
  output2.print("Log on Softmax Output:");

  delta.print("LogSoftMax Derivative:");
  delta2.print("Log on SoftMax Derivative:");

  return 0;
}
```
Output of the above code: 

```
Input
  -0.6871
   0.7898
   0.2011
   0.0949
  -0.0550
LogSoftMax Output:
  -2.4746
  -0.9977
  -1.5864
  -1.6926
  -1.8425
Log on Softmax Output:
  -2.4746
  -0.9977
  -1.5864
  -1.6926
  -1.8425
LogSoftMax Derivative:
   1.0842
   1.3687
   1.2047
   1.1840
   1.1584
Log on SoftMax Derivative:
   0.5790
  -0.8435
  -0.0233
   0.0798
   0.2079

```

Also, derivative of Softmax was verified manually and the derivative calculated using "Log on Softmax" is also verified using pytorch using the following code: 

```
import torch

input = torch.tensor([-0.6871, 0.7898, 0.2011, 0.0949,-0.0550], requires_grad=True)
output = torch.nn.LogSoftmax(dim=0)(input)
obj = output.sum()
output.retain_grad()
obj.backward()
print("Derivative of the input:", input.grad)
```

Output:

`Derivative of the input: tensor([ 0.5790, -0.8435, -0.0233,  0.0798,  0.2079])`

### PR DESCRIPTION
**Pull request by mrdaybird**
Update the Backward method of LogSoftMax  and fix test corresponding to it. Fixes #3468 (See issue for more details)

- [x] Fix derivative
- [x] Fix tests

See this colab file for my explanation and with references for the new derivative : [LogSoftMax](https://colab.research.google.com/drive/1D_hqaXLb7mFhriFHazuEMp_DPYZmf5Ex#scrollTo=vWU2wnCesDOC).
Also to be noted, the new derivative was tested against JacobianTest(corrected version).

### CODE REVIEW

**Comment by shubham1206agra** [comment]
Can you cite the paper? So I can double check this.

**Comment by mrdaybird** [comment]
See this colab file for my explanation and link to paper: [LogSoftMax](https://colab.research.google.com/drive/1D_hqaXLb7mFhriFHazuEMp_DPYZmf5Ex#scrollTo=vWU2wnCesDOC)

**Comment by mrdaybird** [comment]
I have updated the colab file([LogSoftMax](https://colab.research.google.com/drive/1D_hqaXLb7mFhriFHazuEMp_DPYZmf5Ex#scrollTo=vWU2wnCesDOC)) to include a simpler and a more complete calculation of the derivative. Also, an important reference in understanding the derivative of LogSoftMax is understanding how the derivative of Softmax is calculated. For that, you can see the mlpack implementation(great job to whoever wrote) and also [this blog](https://themaverickmeerkat.com/2019-10-23-Softmax/) is a good reference to understand how the softmax derivative is calculated as well as how it is multiplied with the back-propagated derivatives. The derivative in the mlpack is a further simplified form.

File: src/mlpack/methods/ann/layer/log_softmax_impl.hpp [hunk 1]
```diff
@@ -103,7 +103,7 @@ void LogSoftMaxType<MatType>::Backward(
     const MatType& gy,
     MatType& g)
 {
-  g = arma::exp(input) + gy;
+  g = gy - arma::exp(input) % arma::repmat(arma::sum(gy), input.n_rows, 1);
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/log_softmax_impl.hpp:106]: A fascinating observation that was confusing me very much until I realized: these two implementations are the same when (a) `LogSoftMax` is the last layer in the network, and (b) `NegativeLogLikelihood` is the loss function used.

The reason is that with `NegativeLogLikelihood`, `gy` will always be such that only one element takes the value of `-1`, and all other values are `0`.  Thus, `-arma::exp(input) % arma::repmat(arma::sum(gy), input.rows, 1)` is equivalent to `-arma::exp(input) % -1`, which is just `arma::exp(input)`... and we get the original implementation.

So, when I tested this on the `mnist_simple` example using either implementation... the results were exactly the same!

I couldn't find any other examples in the examples repository that used `LogSoftMax` as the last layer but a different loss function than `NegativeLogLikelihood`.

In any case, after using @mrityunjay-tripathi's derivation in #2364, I agree that this is the correct implementation.

**Review by rcurtin — COMMENTED** [review]
Want to add a note to `HISTORY.md` about this?

I found that the tests still pass with both implementations.  Maybe we should add a Jacobian test to `log_softmax.cpp` too?  I tried that, but it fails---I haven't dug into exactly why yet:

```
/**
 * Jacobian log-softmax module test.
 */
TEST_CASE("JacobianLogSoftMaxLayerTest", "[ANNLayerTest]")
{
  for (size_t i = 0; i < 5; ++i)
  {
    const size_t inputElements = RandInt(2, 1000);

    arma::mat input;
    input.set_size(inputElements, 1);
    input.randu();

    LogSoftMax module;
    module.InputDimensions() = std::vector<size_t>({ inputElements });
    module.ComputeOutputDimensions();

    double error = JacobianTest(module, input);
    REQUIRE(error <= 1e-5);
  }
}
```

Sorry it took so long to get to this.  I've been so heads-down on bandicoot that a lot of mlpack stuff has gotten put on the back burner.

File: src/mlpack/methods/ann/layer/log_softmax_impl.hpp (hunk 1 above)
Comment (mrdaybird) [reply src/mlpack/methods/ann/layer/log_softmax_impl.hpp:106]: That's definitely a interesting observation and also a relief knowing that all the models that uses this layer were not incorrectly trained!

**Comment by mrdaybird** [comment]
I will add the Jacobian test after the PR correcting it is merged.

**Comment by mrdaybird** [comment]
@rcurtin I have added the JacobianTest. It is mostly your implementation from your previous comment, because after looking at it, it was probably better than anything I could think of :sweat_smile:. Hope you don't have a problem with that. One slight change is that,  JacobianTest initializes the input matrix, so there's no need to initialize the matrix before passing.

**Comment by rcurtin** [comment]
Sounds good @mrdaybird; thanks for adding it.  I'm going to debug the static code analysis checks job on this PR, so, don't mind the noise...

**Comment by rcurtin** [comment]
@mlpack-jenkins test this please

**Review by rcurtin — APPROVED** [review]
Sweet, got the static code analysis job fixed on the first try. :)

We should add a note to `HISTORY.md`; if you want to do that, feel free, otherwise I'll handle it during merge. :+1:

Thanks for finding and fixing this important issue!

**Comment by rcurtin** [comment]
Forgot to update the history during merge... oops... so I opened #3479.
